java/src/main/native/src/row_conversion.cu

```diff
@@ -750,139 +722,95 @@ copy_validity_to_rows(const size_type num_rows, const size_type num_columns,
-          auto validity_data = __ballot_sync(participation_mask, my_data & dw_mask);
-          // lead thread in each warp writes data
-          auto const validity_write_offset =
-              validity_data_row_length * (relative_row + i) + relative_col / CHAR_BIT;
-          if (threadIdx.x % warp_size == 0) {
-            *reinterpret_cast<int32_t *>(&this_shared_tile[validity_write_offset]) = validity_data;
-          }
+  auto tile = tile_infos[blockIdx.x];
+  auto const num_tile_cols = tile.num_cols();
+  auto const num_tile_rows = tile.num_rows();
+
+  auto const threads_per_warp = warp.size();
+  auto const rows_per_read = cudf::detail::size_in_bits<bitmask_type>();
+
+  auto const num_sections_x = util::div_rounding_up_unsafe(num_tile_cols, threads_per_warp);
+  auto const num_sections_y = util::div_rounding_up_unsafe(num_tile_rows, rows_per_read);
+  auto const validity_data_row_length = util::round_up_unsafe(
+      util::div_rounding_up_unsafe(num_tile_cols, CHAR_BIT), JCUDF_ROW_ALIGNMENT);
+  auto const total_sections = num_sections_x * num_sections_y;
+
+  // the tile is divided into sections. A warp operates on a section at a time.
+  for (int my_section_idx = warp.meta_group_rank(); my_section_idx < total_sections;
+       my_section_idx += warp.meta_group_size()) {
+    // convert to rows and cols
+    auto const section_x = my_section_idx % num_sections_x;
+    auto const section_y = my_section_idx / num_sections_x;
+    auto const relative_col = section_x * threads_per_warp + warp.thread_rank();
+    auto const relative_row = section_y * rows_per_read;
+    auto const absolute_col = relative_col + tile.start_col;
+    auto const absolute_row = relative_row + tile.start_row;
+    auto const participating = absolute_col < num_columns && absolute_row < num_rows;
+    auto const participation_mask = __ballot_sync(0xFFFFFFFF, participating);
+
+    if (participating) {
+      auto my_data = input_nm[absolute_col] != nullptr ?
+                         input_nm[absolute_col][word_index(absolute_row)] :
+                         std::numeric_limits<uint32_t>::max();
+
+      // every thread that is participating in the warp has 4 bytes, but it's column-based
+      // data and we need it in row-based. So we shuffle the bits around with ballot_sync to
+      // make the bytes we actually write.
+      bitmask_type dw_mask = 1;
+      for (int i = 0; i < threads_per_warp && relative_row + i < num_rows; ++i, dw_mask <<= 1) {
+        auto validity_data = __ballot_sync(participation_mask, my_data & dw_mask);
+        // lead thread in each warp writes data
+        auto const validity_write_offset =
+            validity_data_row_length * (relative_row + i) + relative_col / CHAR_BIT;
+        if (warp.thread_rank() == 0) {
+          *reinterpret_cast<bitmask_type *>(&shared_data[validity_write_offset]) = validity_data;
         }
       }
     }
+  }
 
-    // make sure entire tile has finished copy
-    group.sync();
```

Comment:
I wonder if we should add a comment here, to document that syncing earlier would lead to thread divergence.

Reply:
I spoke with Jake about this to try and get some insight as to why this would be a problem. It was confirmed that it could be an issue, but it isn't clear why it was an issue. I can add a comment here.